src/libcore/num/mod.rs

```diff
@@ -2575,6 +2599,53 @@ cross_sign_from_int_impl!(u64, i8, i16, i32, i64, i128, isize);
 cross_sign_from_int_impl!(u128, i8, i16, i32, i64, i128, isize);
 cross_sign_from_int_impl!(usize, i8, i16, i32, i64, i128, isize);
 
+/// The error type returned when a checked floating type conversion fails.
+#[unstable(feature = "try_from", issue = "33417")]
+#[derive(Debug, Copy, Clone)]
+pub struct TryFromFloatError(());
+
+impl TryFromFloatError {
+    #[unstable(feature = "float_error_internals",
+               reason = "available through Error trait and this method should \
+                         not be exposed publicly",
+               issue = "0")]
+    #[doc(hidden)]
+    pub fn __description(&self) -> &str {
+        "out of range floating type conversion attempted"
+    }
+}
+
+#[unstable(feature = "try_from", issue = "33417")]
+impl fmt::Display for TryFromFloatError {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        self.__description().fmt(fmt)
+    }
+}
+
+macro_rules! try_from_float_impl {
+    ($storage:ty, $target:ident, $($source:ty),*) => {$(
+        #[unstable(feature = "try_from", issue = "33417")]
+        impl TryFrom<$source> for $target {
+            type Error = TryFromFloatError;
+
+            fn try_from(u: $source) -> Result<$target, TryFromFloatError> {
+                use $target;
+
+                let min = $target::MIN as $storage;
+                let max = $target::MAX as $storage;
+                if u as $storage < min || u as $storage > max {
```

Comment:
These basically rules out conversion of `float::INFINITY` and `float::NAN`, i.e. `Ok(f64::from(f32::INFINITY))` and `f64::try_from(f32::INFINITY)` will produce different result.